### PR TITLE
BUGFIX: variation coverage restriction

### LINKED ISSUE #5918: CreateNodeVariant over-reaches to the full specialization set and can create an orphan variant
# `CreateNodeVariant` over-reaches to the full specialization set and can create an orphan variant

## Summary

Creating a specialization variant of a node (e.g. translating a node from `de` to `en`) does not materialize the variant only at the requested target dimension space point — it spans the **entire specialization set** of the target origin. If the parent node has been (hard-)removed from one of those covered dimension space points beforehand, `CreateNodeVariant` still succeeds on the write side and emits a `NodeSpecializationVariantWasCreated` event for a dimension space point where the target parent does not exist. The contentGraph projection then fails on catch-up with *"target parent node is missing"* and the projection is stuck in an error state (permanent poison), because there is no write-side validation that rejects the invalid command.

Because catch-up is asynchronous in production, the editor's command "succeeds" and the projection dies silently afterwards — recovery requires manual event surgery.

## Environment

- Reproduced on **`neos/contentrepository-core` 9.1.4** with the Doctrine DBAL content graph adapter.

## Steps to reproduce

The essential condition is a **specialization chain** in *any* dimension: a node whose parent has been removed at a deeper specialization, followed by a `CreateNodeVariant` that targets a shallower point and fans its coverage out into the removed one. It does **not** require multiple dimensions — a single dimension with a three-level chain is enough. Below: the minimal single-dimension reproduction, then the two-dimension variant we actually hit in production (`language` × `market` translation).

### Minimal reproduction — single dimension

One dimension `market` with a three-level chain `general` → `mid` → `deep`. A document `doc` with a child `content`, both at origin `{general}` (covering `general`, `mid`, `deep`).

1. **Hard-remove `doc` at the deepest level** — `RemoveNodeAggregate` for `doc` at `{deep}`. This cascades to the subtree, so both `doc` and `content` are removed from `{deep}`. State is consistent, no error.
2. **Create a variant at the mid level** — `CreateNodeVariant` for `content`, source `{general}`, target `{mid}`. The coverage is the specialization set of `{mid}` = `{mid}`, **`{deep}`** — so `content` is re-materialized into `{deep}`, where the parent `doc` no longer exists.

```gherkin
@contentrepository
Feature: Creating a specialization variant must not poison the contentGraph projection

  Background:
    Given using the following content dimensions:
      | Identifier | Values             | Generalizations         |
      | market     | general, mid, deep | mid->general, deep->mid |
    And using the following node types:
    """yaml
    'Neos.ContentRepository.Testing:Document': []
    """
    And using identifier "default", I define a content repository
    And I am in content repository "default"
    And the command CreateRootWorkspace is executed with payload:
      | Key                | Value           |
      | workspaceName      | "live"          |
      | newContentStreamId | "cs-identifier" |
    And I am in workspace "live" and dimension space point {"market": "general"}
    And the command CreateRootNodeAggregateWithNode is executed with payload:
      | Key             | Value                         |
      | nodeAggregateId | "lady-eleonode-rootford"      |
      | nodeTypeName    | "Neos.ContentRepository:Root" |
    And the following CreateNodeAggregateWithNode commands are executed:
      | nodeAggregateId | nodeTypeName                            | parentNodeAggregateId  | nodeName |
      | doc             | Neos.ContentRepository.Testing:Document | lady-eleonode-rootford | doc      |
      | content         | Neos.ContentRepository.Testing:Document | doc                    | content  |

  Scenario: A variant whose parent was removed at a deeper specialization must not orphan
    # remove the parent at the deepest level (cascades to its child)
    When the command RemoveNodeAggregate is executed with payload:
      | Key                          | Value                |
      | nodeAggregateId              | "doc"                |
      | coveredDimensionSpacePoint   | {"market": "deep"}   |
      | nodeVariantSelectionStrategy | "allSpecializations" |
    # create a variant at the mid level -> coverage fans out to "deep" where doc is gone
    When the command CreateNodeVariant is executed with payload:
      | Key             | Value                 |
      | nodeAggregateId | "content"             |
      | sourceOrigin    | {"market": "general"} |
      | targetOrigin    | {"market": "mid"}     |
    When I am in workspace "live" and dimension space point {"market": "mid"}
    Then I expect node aggregate identifier "content" to lead to node cs-identifier;content;{"market":"mid"}
```

### Two-dimension variant

Two dimensions, `language` (`de` → `en`) and `market` (`consumers` → `businesses`, `busTraffic`); a node translated from `de` to `en` after a market had been pruned. `doc` and `content` are created at origin `{de, consumers}` (covering all six dimension space points via shine-through).

1. **Hard-remove `doc` from one market** — `RemoveNodeAggregate` for `doc` at `{de, busTraffic}` with strategy `allSpecializations`. This cascades to the subtree, so both `doc` and `content` are removed from `{de, busTraffic}` and `{en, busTraffic}`. State is consistent, no error.
2. **Create the English variant of the child** — `CreateNodeVariant` for `content`, source `{de, consumers}`, target `{en, consumers}` (this is what the editor/translation issues when you edit the node in English). The coverage is the specialization set of `{en, consumers}` = `{en, consumers}`, `{en, businesses}`, **`{en, busTraffic}`** — so `content` is re-materialized into `{en, busTraffic}`, where the parent `doc` no longer exists.

```gherkin
@contentrepository
Feature: Creating a language variant must not poison the contentGraph projection

  Background:
    Given using the following content dimensions:
      | Identifier | Values                            | Generalizations                              |
      | language   | de, en                            | en->de                                       |
      | market     | consumers, businesses, busTraffic | businesses->consumers, busTraffic->consumers |
    And using the following node types:
    """yaml
    'Neos.ContentRepository.Testing:Document': []
    """
    And using identifier "default", I define a content repository
    And I am in content repository "default"
    And the command CreateRootWorkspace is executed with payload:
      | Key                | Value           |
      | workspaceName      | "live"          |
      | newContentStreamId | "cs-identifier" |
    And I am in workspace "live" and dimension space point {"language": "de", "market": "consumers"}
    And the command CreateRootNodeAggregateWithNode is executed with payload:
      | Key             | Value                         |
      | nodeAggregateId | "lady-eleonode-rootford"      |
      | nodeTypeName    | "Neos.ContentRepository:Root" |
    And the following CreateNodeAggregateWithNode commands are executed:
      | nodeAggregateId | nodeTypeName                            | parentNodeAggregateId  | nodeName |
      | doc             | Neos.ContentRepository.Testing:Document | lady-eleonode-rootford | doc      |
      | content         | Neos.ContentRepository.Testing:Document | doc                    | content  |

  Scenario: A language variant of content whose parent was removed in an unrelated market succeeds
    When the command RemoveNodeAggregate is executed with payload:
      | Key                          | Value                                      |
      | nodeAggregateId              | "doc"                                      |
      | coveredDimensionSpacePoint   | {"language": "de", "market": "busTraffic"} |
      | nodeVariantSelectionStrategy | "allSpecializations"                       |
    When the command CreateNodeVariant is executed with payload:
      | Key             | Value                                     |
      | nodeAggregateId | "content"                                 |
      | sourceOrigin    | {"language": "de", "market": "consumers"} |
      | targetOrigin    | {"language": "en", "market": "consumers"} |
    When I am in workspace "live" and dimension space point {"language": "en", "market": "consumers"}
    Then I expect node aggregate identifier "content" to lead to node cs-identifier;content;{"language":"en","market":"consumers"}
```

## Expected behaviour

Creating the `{en, consumers}` variant of `content` should only materialize the variant in dimension space points where the parent actually exists

## Actual behaviour

The command is accepted and persisted. The contentGraph projection then throws on catch-up:

```
Failed to create node specialization variant for node "content" ...
because the target parent node "doc" is missing
```

(`ProjectionCatchUpFailed` / `CatchUpHadErrors`; the DBAL adapter throws in `Projection/Feature/NodeVariation.php`, code `1716498734`.) The subscription is left in an error state and stops applying further events.

## Root cause

In `Classes/Feature/Common/NodeVariationInternals.php`, the covered dimension space points for a specialization variant are computed purely from the interdimensional variation graph:

```php
// NodeVariationInternals::calculateEffectiveVisibility / specialization coverage
$this->getInterDimensionalVariationGraph()->getSpecializationSet(
    $targetOrigin->toDimensionSpacePoint(),
    true,        // includeOrigin
    $excludedSet // only excludes points already occupied by another variant
);
```

This spans the full specialization set of the target origin, subtracting only points already occupied by another variant of the same aggregate. It does **not** consider:

- dimension space points from which the node (or its parent) was previously removed, and
- whether the parent aggregate actually covers each target point.

`Classes/Feature/NodeVariation/NodeVariation.php` validates the source parent but never checks parent presence per covered target dimension space point, so an orphan `NodeSpecializationVariantWasCreated` event is emitted and only the (async) projection notices the inconsistency.

### PR DESCRIPTION
**Pull request by nezaniel**
Resolves #5918 

This prefilters the dimension space point set used for variation by evaluating the parent aggregate's coverage

**Review instructions**

Check out the new test case

**Checklist**

- [x] Code follows the PSR-12 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the correct branch: 9.0
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.ContentRepository.Core/Classes/Feature/Common/NodeVariationInternals.php [hunk 1]
```diff
@@ -307,11 +326,14 @@ private function resolveInterdimensionalSiblings(
...
         OriginDimensionSpacePoint $targetOrigin,
-        NodeAggregate $nodeAggregate
+        NodeAggregate $nodeAggregate,
+        ?NodeAggregate $parentNodeAggregateForCoverageFilter,
     ): DimensionSpacePointSet {
         $specializations = $this->getInterDimensionalVariationGraph()
             ->getIndexedSpecializations($targetOrigin->toDimensionSpacePoint());
-        $excludedSet = new DimensionSpacePointSet([]);
+        $excludedSet = $parentNodeAggregateForCoverageFilter
+            ? $specializations->getDifference($parentNodeAggregateForCoverageFilter->coveredDimensionSpacePoints)
```
Comment (mhsdesign) [inline Neos.ContentRepository.Core/Classes/Feature/Common/NodeVariationInternals.php:335]: nitpik:

we pass the full `NodeAggregate` along but we only access `coveredDimensionSpacePoints` as the parameter name suggests. Could be changed to 

```
?DimensionSpacePointSet $parentNodeAggregateCoverageFilter,
```

but can also keep as is;)

File: Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/07-NodeRemoval/06-CreateNodeSpecializationVariantAfterDeletion.feature [hunk 2]
```diff
@@ -259,3 +259,156 @@ Feature: Create node specialization
...
+    And I expect a node identified by cs-identifier;nodimer-tetherton;{"example":"source"} to exist in the content graph
+    And I expect a node identified by cs-identifier;nodimer-tetherton;{"example":"spec"} to exist in the content graph
+    And I expect a node identified by cs-identifier;invariable-mc-nodeface;{"example":"source"} to exist in the content graph
+    And I expect a node identified by cs-identifier;younger-mc-nodeface;{"example":"source"} to exist in the content graph
+    And I expect a node identified by cs-identifier;youngest-mc-nodeface;{"example":"source"} to exist in the content graph
+
+    When I am in workspace "live"
+    Then I expect the node aggregate "lady-eleonode-rootford" to exist
+    And I expect this node aggregate to occupy dimension space points [{}]
+    And I expect this node aggregate to cover dimension space points [{"example":"source"},{"example":"spec"},{"example":"leafSpec"}]
```
Comment (mhsdesign) [inline Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/07-NodeRemoval/06-CreateNodeSpecializationVariantAfterDeletion.feature:305]: i see no other node than the rootnode is in leafSpec as we removed david which was parent to all

File: Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/07-NodeRemoval/06-CreateNodeSpecializationVariantAfterDeletion.feature [hunk 3]
```diff
@@ -259,3 +259,156 @@ Feature: Create node specialization
...
+    And I expect this node aggregate to occupy dimension space points [{"example":"source"}]
+    And I expect this node aggregate to cover dimension space points [{"example":"source"},{"example":"spec"}]
+
+    And I expect the node aggregate "elder-mc-nodeface" to exist
+    And I expect this node aggregate to occupy dimension space points [{"example":"source"}]
+    And I expect this node aggregate to cover dimension space points [{"example":"source"},{"example":"spec"}]
+
+    And I expect the node aggregate "nody-mc-nodeface" to exist
+    And I expect this node aggregate to occupy dimension space points [{"example":"source"},{"example":"spec"}]
+    And I expect this node aggregate to cover dimension space points [{"example":"source"},{"example":"spec"}]
```
Comment (mhsdesign) [inline Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/07-NodeRemoval/06-CreateNodeSpecializationVariantAfterDeletion.feature:321]: the fix of this pr;)

**Review by mhsdesign — APPROVED** [review]
Thank you for the quick fix <3

Is there anything else to consider? Current coverage and node variation of a single node? I guess that would just recreate fallback variants.

Crazy that we missed this earlier :O

Do we need an event migration? Probably only if we have a handful of people having this problem right? Because the projection did die and not corrupt which is good in our case as a replay should just fix this.

File: Neos.ContentRepository.Core/Classes/Feature/Common/NodeVariationInternals.php (hunk 1 above)
Comment (nezaniel) [reply Neos.ContentRepository.Core/Classes/Feature/Common/NodeVariationInternals.php:335]: Yep, the whole thing needs some overhaul, but that's a problem for future us IMHO.
The parent aggregate was already available and this way it's consistent with the rest of the code.

File: Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/07-NodeRemoval/06-CreateNodeSpecializationVariantAfterDeletion.feature (hunk 2 above)
Comment (nezaniel) [reply Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/07-NodeRemoval/06-CreateNodeSpecializationVariantAfterDeletion.feature:305]: Yep

File: Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/07-NodeRemoval/06-CreateNodeSpecializationVariantAfterDeletion.feature (hunk 3 above)
Comment (nezaniel) [reply Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/07-NodeRemoval/06-CreateNodeSpecializationVariantAfterDeletion.feature:321]: Plus we get to this point in the first place

**Comment by nezaniel** [comment]
> Is there anything else to consider? Current coverage and node variation of a single node? I guess that would just recreate fallback variants.

Current coverage is already checked so the same DSP is not covered twice. Apart from that, current coverage is not suitable for filtering as variation is explicitly allowed to create new variants.
Recreation of fallback is not supported yet, that would be a new command (which is in preparation, but for some years now)

> Crazy that we missed this earlier :O

Yep; node variation has had complete edge case analysis - but only before node removal -.-
We definitively need more tests for this, but then again, when don't we (:

> Do we need an event migration? Probably only if we have a handful of people having this problem right? Because the projection did die and not corrupt which is good in our case as a replay should just fix this.

I'd forgo the event migration to get this released asap; also this can only appear once per project as it grinds the cr to a halt. We should add a short description on how to remove the broken event and reactivate the graph projection to the release notes, though

**Review by mhsdesign — APPROVED** [review]
Thanks for the elaboration. Fully makes sense. Merging this so it's already in the up-merge.
